router/usecase/router_usecase_test.go

```diff
@@ -589,6 +589,123 @@ func (s *RouterTestSuite) TestConvertRankedToCandidateRoutes() {
+	// The reason is that there is an alloyed transmuter for routes between allUSDT and kava.USDT
+	// that provides no slippage swaps. Given that 100K is under the liqudiity of kava.USDT in the
+	// transmuter pool, the split routes should be essentially the same.
+	// Update: as of 30.06.24, the kava.usdt for osmo only has one optimal route.
+	const usdtOsmoExpectedRoutesHighLiq = 2
+	var oneHundredThousandUSDValue = osmomath.NewInt(100_000_000_000)
+
+	tests := map[string]struct {
+		tokenInDenom  string
+		tokenOutDenom string
+
+		amountIn osmomath.Int
+
+		expectedRoutesCount int
+	}{
+		// This pair originally caused an error due to the lack of filtering that was
+		// added later.
+		"usdt for umee": {
+			tokenInDenom:  USDT,
+			tokenOutDenom: UMEE,
+
+			amountIn: osmomath.NewInt(1000_000_000),
+
+			expectedRoutesCount: 1,
+		},
+		"uosmo for uion": {
+			tokenInDenom:  UOSMO,
+			tokenOutDenom: UION,
+
+			amountIn: osmomath.NewInt(5000000),
+
+			expectedRoutesCount: 1,
+		},
+		"usdt for atom": {
+			tokenInDenom:  USDT,
+			tokenOutDenom: ATOM,
+
+			amountIn: osmomath.NewInt(5000000),
+
+			expectedRoutesCount: 1,
+		},
+		"uakt for umee": {
+			tokenInDenom:  AKT,
+			tokenOutDenom: UMEE,
+
+			amountIn: osmomath.NewInt(100_000_000),
+
+			expectedRoutesCount: 1,
+		},
+		// This test validates that with a greater max routes value, SQS is able to find
+		// the path from umee to stOsmo
+		"umee for stosmo": {
+			tokenInDenom:  UMEE,
+			tokenOutDenom: stOSMO,
+
+			amountIn: osmomath.NewInt(1_000_000),
+
+			expectedRoutesCount: 1,
+		},
+
+		"atom for akt": {
+			tokenInDenom:  ATOM,
+			tokenOutDenom: AKT,
+
+			amountIn: osmomath.NewInt(1_000_000),
+
+			expectedRoutesCount: 1,
+		},
+
+		"allUSDT for uosmo": {
+			tokenInDenom:  ALLUSDT,
+			tokenOutDenom: UOSMO,
+
+			amountIn: oneHundredThousandUSDValue,
+
+			expectedRoutesCount: usdtOsmoExpectedRoutesHighLiq,
+		},
+		"kava.USDT for uosmo - should have the same routes as allUSDT for uosmo": {
+			tokenInDenom:  ALLUSDT,
+			tokenOutDenom: UOSMO,
+
+			amountIn: oneHundredThousandUSDValue,
+
+			expectedRoutesCount: usdtOsmoExpectedRoutesHighLiq,
+		},
+	}
+
+	for name, tc := range tests {
+		tc := tc
+		s.Run(name, func() {
+			// Setup mainnet router
+			mainnetState := s.SetupMainnetState()
+
+			// Mock router use case.
+			mainnetUseCase := s.SetupRouterAndPoolsUsecase(mainnetState)
+
+			// System under test
+			quote, err := mainnetUseCase.Router.GetOptimalQuoteInGivenOut(context.Background(), sdk.NewCoin(tc.tokenInDenom, tc.amountIn), tc.tokenOutDenom)
+
+			// We only validate that error does not occur without actually validating the quote.
+			s.Require().NoError(err)
+
+			// TODO: update mainnet state and validate the quote for each test stricter.
+			quoteRoutes := quote.GetRoute()
+			s.Require().Len(quoteRoutes, tc.expectedRoutesCount)
+
+			// Validate that the quote is not nil
+			s.Require().NotNil(quote.GetAmountOut())
```

Comment:
I know this was copied form "out given in" but one piece of validation that we could add is that the route is valid.


That is, by following the pools in the returned route, we can assert that the sequence is appropriate and tokens in and out exist in the relevant pools lover the route.

Reply:
In terms of testing route/pools, let's consider that we are extending [TestGetOptimalQuote_Mainnet](https://github.com/osmosis-labs/sqs/blob/e1dcc8c6dc487293c8787e1ba50e4c72b7e48d60/router/usecase/optimized_routes_test.go#L555) instead for now. I may be missing something important here, but it seems that discussed validation will be not possible to implement since GetTokenInDenom is not set for the pools, - it is always empty quotes. That is visible in the [response](https://github.com/osmosis-labs/sqs/blob/e1dcc8c6dc487293c8787e1ba50e4c72b7e48d60/router/usecase/routertesting/parsing/quote_amount_in_response.json#L1) for the clients as well.
Now,  back to the [TestGetOptimalQuoteInGivenOut_Mainnet](https://github.com/osmosis-labs/sqs/blob/e1dcc8c6dc487293c8787e1ba50e4c72b7e48d60/router/usecase/router_usecase_test.go#L594), since GetOptimalQuoteInGivenOut is just a wrapper over quoteExactAmountIn call to GetTokenInDenom still returns empty string. Even if [response](https://github.com/osmosis-labs/sqs/blob/e1dcc8c6dc487293c8787e1ba50e4c72b7e48d60/router/usecase/routertesting/parsing/quote_amount_out_response.json#L1)  contains token_in_denom value we don't have it in the underlying data structures, because it's actually  token_out_denom , - result of inversion we apply during [PrepareResult](https://github.com/osmosis-labs/sqs/blob/e1dcc8c6dc487293c8787e1ba50e4c72b7e48d60/router/usecase/quote_out_given_in.go#L65) call. Note that this method is already [covered](https://github.com/osmosis-labs/sqs/blob/e1dcc8c6dc487293c8787e1ba50e4c72b7e48d60/router/usecase/quote_test.go#L40) with tests.
I suspect this the reason why other tests such as [ValidateRouterPools](https://github.com/osmosis-labs/sqs/blob/e1dcc8c6dc487293c8787e1ba50e4c72b7e48d60/router/usecase/routertesting/suite.go#L308) does not implement such test coverage either.  Please let me know whether I am missing something here or it's actually not feasible at this point?